### ISSUE TITLE
Add `osxphotos exportdb --history`

### ISSUE THREAD

**Issue by RhetTbull** [RhetTbull/osxphotos#1311]
`osxphotos exportdb --history FILE_OR_UUID`  would print out history of a given file or URL -- for example, the date and what happened (exported, skipped, updated + reason)

**Comment by RhetTbull** [comment]
@oPromessa yes, that's one of the intended use cases. Ideally, you'd be able to see select an exported file (or a UUID) and see the entire history of when it was exported, skipped, updated, etc. and if updated, what the reason was. Without blowing up the export database to save a lot more data, I can't show exactly what changed (for example, a new label was added by photoanalysisd) only that "the metadata changed" or "the touch time changed", etc. But this should useful determining why photos are getting updated.  For example, currently, the exporter uses the following reasons for determining if a photo should be updated. My idea is to store this reason so at least you know why the photo was updated.

```python
class ShouldUpdate(Enum):
    NOT_IN_DATABASE = 1
    HARDLINK_DIFFERENT_FILES = 2
    NOT_HARDLINK_SAME_FILES = 3
    DEST_SIG_DIFFERENT = 4
    EXPORT_OPTIONS_DIFFERENT = 5
    EXIFTOOL_DIFFERENT = 6
    EDITED_SIG_DIFFERENT = 7
    DIGEST_DIFFERENT = 8
    UPDATE_ERRORS = 9
```

It might be possible to store a diff of the export info for each export to show exactly which label, keyword, etc. changed but that's not part of my current design.  I'll think about this.

**Comment by RhetTbull** [comment]
I recalled adding data about metadata changes to the report and found that I'd added a `metadata_changed` field to the report class in #621 but I'd never added to the actual ReportWriter so the data is stored in the export database (last 10 reports are stored whether or not you used the `--report` flag) but not reported in the actual report file.

I wrote a [quick script](https://github.com/RhetTbull/osxphotos/blob/main/examples/metadata_change_report.py) that will extract the `metadata_changed` field for the last export to tell you what files changed (but not which metadata) and this might be helpful in your case. If the files being updated don't show as having changed metadata, that's a mystery we need to investigate.

If you are running osxphotos v0.65.0 or later, you can run the example code directly from github:

`osxphotos run https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/metadata_change_report.py /path/to/export/folder /path/to/export/database`

Will print out something like this:

```
Export date: 2023-12-05T08:56:17.496995
/Users/user/Desktop/export/IMG_4547.jpg (3DD2C897-F19E-4CA6-8C22-B027D5A71907
```

**Comment by RhetTbull** [comment]
It looks like I can compute a minimal diff for the photo metadata and possibly store that in the export db which would allow you to see exactly what changed between the last two exports:

```python
[
    ("add", "albums", [(4, "New Album")]),
    ("add", "labels", [(4, "New Label")]),
    ("change", "latitude", (-34.91889167000001, -34)),
]
```

**Comment by RhetTbull** [comment]
So the [dictdiffer](https://github.com/inveniosoftware/dictdiffer) library I was looking at using doesn't appear to work with this use case. It has trouble with datetimes, unsorted lists, and sets. I'll probably just roll my own.  Would like to get something like this that can be stored to the exportdb:

```python
[
   ("albums", "add", ["New Album"]),
   ("labels", "add", ["New Label1", "NewLabel2"]),
   ("keywords", "remove", ["vacation"])
]
```

I think this would make it easy to see changes. One downside is that the history would need to be stored for every single action for a photo for the life of the export set. Thus every time you use `--update`, a new entry for every file is created. This could cause the exportdb to grow quite large. Of course, most of the time there are no changes so only a timestamp and `null` are stored, not the diff.  Need to think about this.  Maybe this is an option the user has to turn on?  Maybe add a `--prune-history` option to `osxphotos exportdb` so you can prune the history if needed?

**Comment by oPromessa** [comment]
> If you are running osxphotos v0.65.0 or later, you can run the example code directly from github:
> 
> `osxphotos run https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/metadata_change_report.py /path/to/export/folder /path/to/export/database`
> 

Tested it out. Weird result. 
* Had a run with new pics 292 and 1583 updated...
* followed by another run with 930 updated. But the query on metadata changes only identifies 6. And they are pics loaded way back when!

```sh
$ grep Processed runOSxPhotos.cron.err 
2023-12-03 07:24:23.558682 -- Processed: 94097 photos, exported: 0, updated: 1282, skipped: 92825, updated EXIF data: 1282, missing: 0, error: 0, touched date: 1282
2023-12-05 19:10:39.719176 -- Processed: 94389 photos, exported: 292, updated: 1583, skipped: 92524, updated EXIF data: 1875, missing: 0, error: 0, touched date: 1875
2023-12-08 07:22:12.174845 -- Processed: 94389 photos, exported: 0, updated: 930, skipped: 93469, updated EXIF data: 930, missing: 0, error: 0, touched date: 930
$ osxphotos run https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/metadata_change_report.py /Volumes/photo/Family.Photos /Users/Shared/Pictures/Logs/.osxphotos_export.db  | wc -l
       6
$
```

**Comment by oPromessa** [comment]
A bit more digging. Found a pic which is updated 9 out of the last 20 runs or so. 

As I export to NAS was able to retrieve the previous version. Comparing EXIF data aside from MacOS entries that differ I found the Description field where I put the albums the pic belongs to show the same albums but ordered differently. 

Btw one of the albums I don't export as I filter it. 

Could it be that osxphotos while building and comparing metadata gets the albums sorted differently on each run into the description field?

**Comment by RhetTbull** [comment]
> Could it be that osxphotos while building and comparing metadata gets the albums sorted differently on each run into the description field?

Ahah! That's it! Because of the way the data is loaded from the database the order of multi-value items like keywords and albums is not guaranteed to be consistent as you've discovered.  One solution is to sort them before passing into the template system but I need to think about this to ensure there's no 2nd order effects of doing so. 

The osxphotos template language does provide a `|sort` filter which sorts a list of values so that could be helpful. I'll think about this some more because it was not intuitive that what you're doing should cause the order to change.

Until I get a fix in place, you can sort the output yourself to ensure albums are always in sorted order (this could change a lot of your photos though):

`Album:{album|sort,}{newline}Description:{descr,}`

I'm not sure this template does what you want though. A couple issues: 

- If there's no album it still prints `Album:`
- If there's no description, it still prints `Descr:`
- Most importantly, it prints the description as many times as there are albums because the entire template gets evaluated for the combination of all possible values. Thus, a photo in two albums will get a description like this:

```
  Album:Album 1
  Description:The description
  Album:Album 2
  Description:The description
```

Here's a different template that prints the `Album:` part first (and sorted) then the description. This is more complex but might be a better fit for what you're trying to do:

`{var:NL,{newline}}{var:AlbumDescr,{album?Album:{album|sort,},}}{%AlbumDescr?{%AlbumDescr|join(%NL)},}{descr?{newline}Descr:{descr},}`

This prints out:

```
  Album:Album 1
  Album:Album 2
  Description:The description
```

`{var:NL,{newline}}`: Use `var` variable feature to create a variable called `NL` which holds a newline (needed because templates cannot be used in filter fields but variables can)

`{var:AlbumDescr,{album?Album:{album|sort,},}}`: and a variable called `AlbumDescr` which is the `Album:Album Name` part except that it sorts the albums and only does this if there are albums. If you wanted `Album:` to show up even if no album, you can set the default value to `Album:`

`{%AlbumDescr?{%AlbumDescr|join(%NL)},}`: Test to see if the `%AlbumDescr` variable has a value, if so, use the `|join()` filter to join the values with a `{newline}` but since filters cannot accept templates as arguments, use the `%NL` variable we defined earlier.  The output of this is a single string in form:

```
  Album:Album 1
  Album:Album 2
```

`{descr?{newline}Descr:{descr},}`: Test to see if `{descr}` has a value if so, add a new line then add the `Descr:description` part. 

This adds a new line before description even when there's no album -- out of time at the moment but can go back and tweak later to fix this.

**Comment by RhetTbull** [comment]
@oPromessa looking at the code I see some values like keywords and labels are sorted but others like albums are not. This is a bug and I've opend #1317 to fix this.

**Comment by RhetTbull** [comment]
@oPromessa v0.66.0 now sorts `{album}` and all other "list of string" templates where it makes sense to do so.  This will cause a change to any sidecars and possible `--exiftool` processed photos when run with `--update` but will fix the problem you've been having with updated photos that shouldn't be updated.

**Comment by oPromessa** [comment]
Thanks so much. Testing it out.

Still using:
```
description_template = "Album:{album,}{newline}Description:{descr,}"
directory = "{album|filter(startswith 1|2)[/,.|:,.]}"
```
1. On your notes:
   * If there's no album it still prints Album: => Ok. That's fine for now. Will test it out further your other options
   * If there's no description, it still prints Descr: =>  Ok. That's fine for now. Will test it out further your other options
   * Most importantly, it prints the description as many times as there are albums because the entire template gets evaluated for the combination of all possible values. Thus, a photo in two albums will get a description like this: (...) => I'm not getting duplicated lines for Album. Maybe because:
      * actually only one of the Albums gets exported. The other are control Albums like Duplicated, etc. I (try) not to ever have a pics into two exported Albums.
      * most pics don't actually have a Description.
2. But then I started testing adding a Description in Photos on the Caption field and noticed a different mapping of the fields with exiftool
   * Without a caption in Photos:
```
[IFD0]          ImageDescription                : Album:2016-08-15 XXX.Description:
[IPTC]          Caption-Abstract                : Album:2016-08-15 XXX.Description:
[XMP-dc]        Description                     : Album:2016-08-15 XXX.Description:
```
   * With a caption in Photos... it seems  the field *XMP-dc* is not populated.
```
[IFD0]          ImageDescription                : Album:2012-04-15 YYY, Duplicate Photos.Description:a caption in Photos
[IPTC]          Caption-Abstract                : Album:2012-04-15 YYY, Duplicate Photos.Description:a caption in Photos
```

**Comment by oPromessa** [comment]
Note if I export from Photos it does populate the three Description fields with the info.

Sorry I missed to notice that I got an exiftool's warning pertaining to XMP Region list... which I though not to be relevant as it was not related to Description but now I'm thinking twice. (Found this old reference #350 but possibly not related.)

```
exiftool warning for file /Users/Shared/Pictures/Apps/X/2012-BBBBB/IMG_0315.JPG: Warning: [Minor] Duplicate XMP property: mwg-rs:Regions/mwg-rs:RegionList/rdf:Bag/rdf:li 10/mwg-rs:Type - /var/folders/28/x8vbdy4j4vn680b5pt33hs0w0000gn/T/osxphotos_export_x_5sdwuk/IMG_0315_exiftool.JPG
```

Tried then with -v3 option to exiftool:
```
osxphotos export X --name IMG_0315 --name IMG_20230418_160854  --verbose --verbose --verbose  --load-config colibri.toml --exiftool-option -v3
```
* BTW when adding `-v3`(to show additional output to `exiftool` ... osxphotos does not break the lines output correctly :(
* Replace | for Newlines and it becomes (more) readable ;)

But did not find anything relevant. Tried then export the  sidecar exiftool format and run -tagsfrom file and same Duplicate XMP error. 
```
exiftool -tagsfromfile IMG_0315.json  -all:all IMG_0315.JPG   -v3
(...)
Warning: [Minor] Duplicate XMP property: mwg-rs:Regions/mwg-rs:RegionList/rdf:Bag/rdf:li 10/mwg-rs:Type - IMG_0315.JPG
```
Also trying to set XMP:Description=hello directly on the file, same error and no file change.
```
exiftool -XMP:Description=hello IMG_0315.JPG
Warning: [Minor] Duplicate XMP property: mwg-rs:Regions/mwg-rs:RegionList/rdf:Bag/rdf:li 10/mwg-rs:Type - IMG_0315.JPG
    0 image files updated
    1 image files unchanged
```
Went also the the original JPG file and I get the same error.

In a nutshell this JPG pic from an iPhone 4S is somehow corrupt.
Photos does the job when exporting but maybe he rewrites everything!


PS: Sorry for the long message... didn't want to leave you hanging!

**Comment by oPromessa** [comment]
Excellent results with sorted on Albums! Decreased by 1000's the pics regularly updating.

Now the cases are normal `labels` (Photos updates on its analysis) changes into `keywords`

Super thanks @RhetTbull  for the support!

**Comment by RhetTbull** [comment]
The history feature is close to being done. It will work on either a file path or a UUID:

```
❯ osxphotos exportdb ~/Desktop/export --history ~/Desktop/export/Pumkins2.jpg
2023-12-21 22:11:43.391, export, None
2023-12-21 22:21:46.971, skip, [["keywords", "added", ["Kids"]], ["keywords", "removed", ["Travel"]]]
2023-12-21 22:37:40.478, skip, None
2023-12-22 05:00:16.452, update: EXPORT_OPTIONS_DIFFERENT, None
```

```
> osxphotos exportdb ~/Desktop/export --history F12384F6-CD17-4151-ACBA-AE0E3688539E
2023-12-21 22:11:42.855, export, None
2023-12-21 22:21:47.009, skip, [["keywords", "added", ["Foo"]], ["keywords", "removed", ["Kids"]]]
2023-12-21 22:37:40.354, skip, None
2023-12-22 05:00:16.819, update: EXPORT_OPTIONS_DIFFERENT, None
```

**Comment by oPromessa** [comment]
Just FYI... Using --history one can tell Photos is **confused** adding and removing the same label from time to time. As I'm using label in keywords ;)... osxphotos WORKS!

```python
SomeAlbum/SomePicJPG
2023-12-25 12:41:16.166, skip, None
2023-12-25 23:15:46.728, skip, None
2023-12-26 22:17:43.350, update: EXIFTOOL_DIFFERENT, [["labels", "added", ["Plant"]]]
2023-12-29 07:23:40.951, update: EXIFTOOL_DIFFERENT, [["labels", "removed", ["Plant"]]]
2023-12-30 06:39:35.841, update: EXIFTOOL_DIFFERENT, [["labels", "added", ["Plant"]]]
2023-12-31 04:23:06.291, skip, None
2024-01-01 03:24:31.686, skip, None
2024-01-02 01:11:32.476, skip, None
2024-01-02 20:29:06.708, skip, None
2024-01-05 05:54:12.750, skip, None
2024-01-06 04:54:03.162, update: EXIFTOOL_DIFFERENT, [["labels", "removed", ["Plant"]]]
2024-01-07 04:58:38.396, skip, None
2024-01-07 21:52:24.866, update: EXIFTOOL_DIFFERENT, [["labels", "added", ["Plant"]]]
2024-01-12 04:02:52.074, update: EXIFTOOL_DIFFERENT, [["labels", "removed", ["Plant"]]]
2024-01-13 05:16:36.060, update: EXIFTOOL_DIFFERENT, [["labels", "added", ["Plant"]]]
2024-01-14 03:46:26.685, update: EXIFTOOL_DIFFERENT, [["labels", "removed", ["Plant"]]]
2024-01-15 21:24:12.690, update: EXIFTOOL_DIFFERENT, None
2024-01-18 01:57:30.016, update: EXIFTOOL_DIFFERENT, [["labels", "added", ["Plant"]]]
2024-01-19 06:02:44.650, update: EXIFTOOL_DIFFERENT, [["labels", "removed", ["Plant"]]]
2024-01-20 05:55:39.241, skip, None
2024-01-21 06:19:30.219, skip, None
2024-01-21 17:15:55.205, skip, None
2024-01-26 07:12:14.316, update: EXIFTOOL_DIFFERENT, [["labels", "added", ["Plant"]]]
2024-01-27 04:41:33.635, update: EXIFTOOL_DIFFERENT, [["labels", "removed", ["Plant"]]]
```

A quick question..on labels and face changes the EXIFTOOL_DIFFERENT details is quite superb. But sometimes is simply:

```python
SomAlbum/SomePic.jpg
2023-12-25 12:47:49.001, skip, None
(...)
2024-01-19 07:32:34.463, skip, None
2024-01-20 03:53:06.878, update: EXIFTOOL_DIFFERENT, None
2024-01-21 06:40:33.867, update: EXIFTOOL_DIFFERENT, None
2024-01-21 19:06:43.458, skip, None
2024-01-26 06:06:45.585, skip, None
2024-01-27 03:47:16.458, skip, None
```
I'm at 0.67.3 and maybe previous versions did not have/show more detail?

**Comment by RhetTbull** [comment]
> I'm at 0.67.3 and maybe previous versions did not have/show more detail?

The history feature was added in v0.67.0 so there won't be history data from before that. 

I'm history diff is computed based on information stored in the database that's always been there so I don't think this is the cause of `None` as the diff. I'll take a look to see what would make this happen.

**Comment by RhetTbull** [comment]
So looks like two things could result in `update: EXIFTOOL_DIFFERENT, None`:

1. The photo has no `PhotoInfo` record stored in the export database. This is written for every photo exported and has been for a long time so that's unlikely.
2. There is no difference in the `PhotoInfo` metadata records but there is a difference in the exiftool data that would be written. Looking at the code it might trigger this if the order of data changed (e.g. "Plants", "Flowers" -> "Flowers", "Plants") but the data itself did not change. I can add code to ensure the exiftool data is always sorted before comparison (though it should be as the data in the PhotoInfo record is sorted...so will need to explore).
